mdetsims/metacal/fofs.py

```diff
@@ -0,0 +1,294 @@
+    """Generate FoF groups.
+
+    Parameters
+    ----------
+    cat: array with fields
+        Usually the cat from a meds file
+    fof_conf: dict
+        configuration for the FoF group finder
+    mask: mask object
+        e.g. a star mask.  Objects that are masked are put into
+        their own FoF group, and are not allowed to be part
+        of a group with other objects
+    """
+
+    if mask is not None:
+        is_masked = mask.is_masked(cat['ra'], cat['dec'])
+    else:
+        is_masked = None
+
+    mn = MEDSNbrs(
+        cat,
+        fof_conf,
+        is_masked=is_masked,
+    )
+
+    nbr_data = mn.get_nbrs()
+
+    nf = NbrsFoF(nbr_data)
+    fofs = nf.get_fofs()
+
+    return nbr_data, fofs
+
+
+def make_singleton_fofs(cat):
+    """Generate a fofs file, one object per groups
+
+    Parameters
+    ----------
+    cat: array with fields
+        Should have a 'number' entry
+
+    Returns
+    -------
+    Fof group array with fields, entries 'fofid', 'number'
+    """
+    dt = [('fofid', 'i8'), ('number', 'i8')]
+    fofs = np.zeros(cat.size, dtype=dt)
+    fofs['fofid'] = np.arange(fofs.size)
+    fofs['number'] = cat['number']
+    return fofs
+
+
+class MEDSNbrs(object):
+    """Gets nbrs of any postage stamp in the MEDS.
+
+    A nbr is defined as any stamp which overlaps the stamp under consideration
+    given a buffer or is in the seg map. See the code below.
+
+    Options:
+        buff_type - how to compute buffer length for stamp overlap
+            'min': minimum of two stamps
+            'max': max of two stamps
+            'tot': sum of two stamps
+
+        buff_frac - fraction by whch to multiply the buffer
+
+        maxsize_to_replace - postage stamp size to replace with maxsize
+        maxsize - size ot use instead of maxsize_to_replace to compute overlap
+
+        check_seg - use object's seg map to get nbrs in addition to postage
+                    stamp overlap
+    """
+    def __init__(self, meds, conf, is_masked=None):
+        self.meds = meds
+        self.conf = conf
+
+        if is_masked is None:
+            is_masked = np.zeros(meds.size, dtype=bool)
+        self.is_masked = is_masked
+        self.is_unmasked = ~is_masked
+
+        self._init_bounds()
+
+    def _init_bounds(self):
+        if self.conf['method'] == 'radius':
+            return self._init_bounds_by_radius()
+        else:
+            raise NotImplementedError(
+                'stamps not implemented for ra,dec version')
+            return self._init_bounds_by_stamps()
+
+    def _init_bounds_by_radius(self):
+
+        radius_name = self.conf['radius_column']
+
+        min_radius = self.conf.get('min_radius_arcsec', None)
+        if min_radius is None:
+            # arcsec
+            min_radius = 1.0
+
+        max_radius = self.conf.get('max_radius_arcsec', None)
+        if max_radius is None:
+            max_radius = np.inf
+
+        m = self.meds
+
+        # switch to row, col MRB
```

Comment:
Below here I switched some of the logic around the boundaries.

Reply:
looks ok